### DISCUSSION TITLE
opentrack 2021.2.0

### DISCUSSION THREAD

**Discussion by sthalik** [opentrack/opentrack#1381] [Announcements]
### Notes

Workaround Windows Defender false positive by compiling the release with Visual Studio 2022 (#1371).

### Features

Add another slider to the PT point filter. Setting it to a low value prevents sawtooth-like motion. The point filter greatly reduces the need for using Accela or any other generic filter.

<hr /><em>This discussion was created from the release <a href='https://github.com/opentrack/opentrack/releases/tag/opentrack-2021.2.0'>opentrack 2021.2.0</a>.</em>

**Comment by Len62** [comment]
All is fine here with this version, no false positives. Have to play with the new PT point filter some tomorrow, thanks for the new update!

**Reply by vdruts** [reply]
I'm using Acuro could you please add the PT point filter? I find Acuro to be superior than most methods.

**Comment by Juanillus** [comment]
I have updated from 1.2 version and in my case the 2.0 version has a terrible jittering in DCS that makes it unplayable. Please, could you put a link to the 1.2 version to reinstall the olde version again?

**Comment by sthalik** [comment]
2021.1.3 is below. There weren't any changes that affect DCS though.

**Reply by Juanillus** [reply]
I tried 2021.1.3 also, and I had the same jittering issue with both 2021.2.0 and 2021.1.3.  But 2021.1.2 worked fine.
I solved the jittering by downgrading to version 2021.1.0 that I found on other page. It could be some change from 2021.1.2 to 2021.1.3.
Thank for your response.

**Comment by sthalik** [comment]
That might be related to a change in the PT tracker. Basically, you need to adjust the threshold and min/max point size as the computed point size is different. So drag the slider a bit more to the right.

**Reply by Juanillus** [reply]
By the moment I'll stick to the older version, it works very well. Just if it can help your develope, in the newer version the extraction blue circles were shaking and slightly "ellipctic". Thank you.

**Comment by sthalik** [comment]
The points are the same for tracking purposes. It's only the visualization that's changed.

**Comment by 3itZ** [comment]
> I tried 2021.1.3 also, and I had the same jittering issue with both 2021.2.0 and 2021.1.3. But 2021.1.2 worked fine. I solved the jittering by downgrading to version 2021.1.0 that I found on other page. It could be some change from 2021.1.2 to 2021.1.3. Thank for your response.

Absolutely the same problem. I stopped at version 2021.1.2, which works correctly.

**Comment by sthalik** [comment]
Have you adjusted the target point size in the PT tracker with the slider already? It needs to be larger by a factor of 1.75 compared to 2021.1.2.

**Comment by 3itZ** [comment]
> Have you adjusted the target point size in the PT tracker with the slider already? It needs to be larger by a factor of 1.75 compared to 2021.1.2.

Before writing about my problem, I tried everything. Dot size, camera settings, point filter settings, etc. I fiddled with this for several evenings...

**Comment by WF000** [comment]
This is how i do it and is rock solid, well in BMS , no DCS yet.

![Screenshot 2021-12-09 111344](https://user-images.githubusercontent.com/20553377/145377652-ec90bdc2-7f60-4dbb-879d-f24738b9b7d1.jpg)

**Reply by Len62** [reply]
Ok, i set-up using your configuration settings and it works nicely. I'll fire-up a game shortly for further testing and report back. Thanks WF000!

**Reply by Rayspit** [reply]
I tried to download Smoothtrack on my Android phone and it says that it's not compatible with my phone. I have Galaxy M12 phone. Thanks in advance

**Comment by Len62** [comment]
Ok, I fired up IL2 Great Battles and I must say using these settings has made movement much smoother and more fluid. It would seem that no filter with point filter checked makes quite a difference.

**Comment by WF000** [comment]
No prob.  Well, 2cc , .. after a longer period I noticed a bit of jitter down -Z axis (zoom, translation forward/back).

Accela is still superior then Point filter as no any jitter, never... so I reverted to Accela...

One BUG for this version tho, Halik.

When output set as Freetrack  ,  if selection is NOT on "Enable Both" (trackir and freetrack) , then it deletes the path to Naturalpoint reg entry.
So, no reg entry to ../modules/  if not set to both. 
... of course TrackIR support is not working then. 
I prefer freetrack disabled (only trackir) as no collision possible then.
![Screenshot 2021-12-09 175335](https://user-images.githubusercontent.com/20553377/145441216-2a027171-2c31-4e2a-9927-7278fa30e7fc.jpg)

Cheers

**Comment by dglava** [comment]
2021.2.0 is giving me really bad results. I am using the same settings (same profile, same filter, no PT filter) and the movement seems erratic and jumpy. It seems more sensitive and movement seems to involve unused axes as well (looking to the right moves it up an down). Have there been any changes which would explain that?

I tried the same settings as WF000, but with the same issues. On video it's subtle, but it's enough to make playing unpleasant.

Here's a video of 2021.2.0:
https://i.imgur.com/FkUel9d.mp4

Here's a video of 2.3.13:
https://i.imgur.com/CmAzQ0l.mp4

**Comment by sthalik** [comment]
@dglava confirm if this build helps. https://www.dropbox.com/s/6wre9q8wzawpp80/opentrack-test-20211214_01.7z

Otherwise I'm gonna have to revert more things.

**Reply by dglava** [reply]
Barely any change. I have a video of using it with the PT filter and without it if it helps:
https://i.imgur.com/kJlywLM.mp4

I have also tested version 2021.1.3 with the same results. Going one version back to 2.3.13 makes it behave in an acceptable way. I am not sure if the issue is due to my settings or tracking device though and I understand that it's not a good "bug report".

**Comment by Vortex76** [comment]
I have the same problem, jumpiness mostly noticable on yaw axis ( where movement is longer usually ) like its lost the smoothness it had on 2.3.12. I tried all settings proposed above and also the build you gave to dglava but it remains the same. Its like its tracking fine, but accella loses its charm, and at least for me I cant go further than 2.3.12, heck I just made an account here so I can write about it. Please help us. You rule!

**Comment by sthalik** [comment]
I need to know your setups.

1. Are you using the ps3eye open driver?
1. Does it help to switch to/from the open driver?
1. Are you using the point filter?
1. Is the jumpiness still present when using the Hamilton filter?

**Reply by Vortex76** [reply]
1. I use open driver a couple of months now. 
2. No problems with or without it. A bit faster and more responsive when on it.
3. Yes and jumpiness on 2.0 - No and jumpiness on 2.0 - No and all good on 1.2.
4. Tried with Hamilton but I cant get it to stabilize. Is even worst than when on Accela or PF.
5. If it helps the jumpiness doesnt show in the octopus monitor. It is there in game ( DCS ).

**Reply by sthalik** [reply]
See if this https://github.com/opentrack/opentrack/releases/tag/opentrack-2021.1.2 version works. I believe that the problem was introduced in 2021.1.3.

**Reply by Vortex76** [reply]
Yes thats the last one that was butter smooth. I hope you nail it! We always have this one to roll  back to, but we also appreciate your efforts to bring more goodies and new versions to us. Thanks in advance!

**Comment by sthalik** [comment]
So let's bisect this. I'm going to have to send you a couple of builds, but by the time we're done it's going to be evident as to which commit broke filtering.

https://www.dropbox.com/s/kxlfpcrvuhqr8ns/opentrack-bisect_01.tar.xz

**Reply by Vortex76** [reply]
This one where I suppose only pf is introduced has no problems. It is smooth. With pf enabled and without it. But I believe something changed in threshold behavior too. I cant grasp what it is but its "useful layer/sweetspot" limits somehow got narrower. Before that there was a wider area where it could let tracking work, now you have to find the exact spot. I m not messing with min/max I keep it where it was at min 1px max 20px. Radius sweetspot 8.92

**Reply by sthalik** [reply]
Bisecting: 26 revisions left to test after this (roughly 5 steps)
https://www.dropbox.com/s/ab0po74khg8xthk/opentrack-bisect_02.tar.xz

> something changed in threshold behavior too. I cant grasp what it is but its "useful layer/sweetspot" limits somehow got narrower. Before that there was a wider area where it could let tracking work, now you have to find the exact spot. 

I widened the scale. Now it's suitable with a larger range of devices. There's a hardware-based tracker in the works that needs values outside the previous range.

**Reply by Vortex76** [reply]
02 Very smooth too no jumpiness at all. Also Threshold works as expected maybe even better than before.

**Reply by sthalik** [reply]
Bisecting: 13 revisions left to test after this (roughly 4 steps)
[848e414dbf0404b5ad3b97a00a06cd0d7dd8c95c] tracker/pt: fix hang

https://www.dropbox.com/s/ly6vr376n0b30zs/opentrack-bisect_03.tar.xz

**Reply by Vortex76** [reply]
03 Not really jumpy, but movement on pitch/yaw became very sensitive and when you turn pf on its even more, especially noticeable on pitch where range is smaller. I believe this might be part, or the start of the problem and maybe exaggerated  more on later revisions where it starts to seem like skipping.

**Reply by sthalik** [reply]
Bisecting: 6 revisions left to test after this (roughly 3 steps)
[f6f9c8a7694dce77fe03c9ea49a3b37bd365b9c6] clang-tidy: update
https://www.dropbox.com/s/6ocrj5987g9bk9n/opentrack-bisect_04.tar.xz

> movement on pitch/yaw became very sensitive and when you turn pf on its even more

There shouldn't be any such difference between 02 and 03. Marking as good, moving on.

**Reply by Vortex76** [reply]
04 Pretty stable too. No problems.

**Reply by sthalik** [reply]
Bisecting: 3 revisions left to test after this (roughly 2 steps)
[471a4e360d262cf3de65cb107eab477d7765981d] dinput: disable joystick input on msvc+asan
https://www.dropbox.com/s/4uq9hjigquxij6q/opentrack-bisect_05.tar.xz

**Reply by Vortex76** [reply]
05 Is jumpy and also movement on yaw outputs in pitch axis too as far as I can see. Gives out a very confusing outcome. Also has different feel with pf on and off.

**Reply by sthalik** [reply]
Good. So it's 76d9145b and indeed it's the point filter that's the problem.

**Reply by dglava** [reply]
I have tested the bisected builds and I'm getting the bad behaviour already on bisect_01.
2021.1.2 works as expected.

>Are you using the ps3eye open driver?

I am using the ps3eye open driver.

>Are you using the point filter?

With the filter it appears smoother, but it's still jumpy.

>Is the jumpiness still present when using the Hamilton filter?

I don't have Hamilton available in bisect_01, but I am generally using Hamilton and experienced the jumpiness with the newer versions with it.

**Reply by sthalik** [reply]
@dglava can you message sthalik on irc.rizon.net, sthalik on skype, or join the discord at <<https://discord.com/invite/NBVX6arwnv>>?

**Reply by sthalik** [reply]
@dglava sent you a PM on Discord.

**Comment by Len62** [comment]
Sorry I can't join in on the testing this morning, will try later on today.

**Comment by sthalik** [comment]
@Vortex76 please confirm whether this helps, or at least changes something at all. https://www.dropbox.com/s/a66rrn2j7na7usx/opentrack-pf-test_01.tar.xz

**Reply by Vortex76** [reply]
Yes this is much better. Still not like 2.3.12 though.  
Then, Pf behavior feels more logarithmic and rounded, but it introduces some lag so it makes me wanna have it off.

**Reply by sthalik** [reply]
I'm curious. What happens if you switch PT color mode to red only or green only? Is it any different than natural/average? I'm asking because if anything else than natural/avg is selected, then RGB mode is used with the ps3eye driver.

**Reply by Vortex76** [reply]
No difference with red green or blue. Feels same like nat/avg. If switching to any of the chroma keys it doesn't track at all.
But I m on Delan Clip which uses infrared leds so maybe others might notice different behavior.

**Comment by Len62** [comment]
@sthalik Ok, finally had a chance to look at 2021.2.0. Using accela filter operation is normal as far as lag, jittery movement. Turning accela off and using the new PF, operation is normal with the exception that movement tends to drift slightly, could be just a setting that needs adjusting in the PF settings. Using accela, if I turn my head I can stop at any point and everything stays steady, with the PF filter position tends to drift by a few degrees. I sometimes wonder if the jittery movement can be attributed to the games themselves and how new scenery gets drawn and compiled in memory. I haven't fired-up DCS recently, my experiences are with IL2 Great Battles.

**Reply by Len62** [reply]
Ok, tested in DCS World and with the accela filter enabled it was fine. When I switched to using the PF setting stuttering was more evident. Have to play around with PF settings to see if it any improvement is noticed.

**Reply by sthalik** [reply]
You need to lower Accela's deadzone to make the pf stutter go away. Normal pf settings are around 1.1 with limit set to 0.2 or slightly lower.

**Reply by Len62** [reply]
So, use accela and the PF at the same time? I was not selecting a filter when using the point filter.

**Reply by sthalik** [reply]
Yup. Use the low intensity point filter to stabilize the view for lower Accela smoothing, i.e. first slider. You can keep the PF setting at some 1.05, 0.2 and overall gain a faster response than with Accela alone.

**Reply by Len62** [reply]
Ok, here are the settings I used: 
![image](https://user-images.githubusercontent.com/10486813/146681258-cc2203fe-ec9a-40f5-9dea-f03c6e830214.png)
![image](https://user-images.githubusercontent.com/10486813/146681265-20ef68b9-2803-4f82-8a9c-3b087300690c.png)
Things seem to be working really well in both IL2 and DCS. Will experiment some more over the next few days.

**Comment by animusVoxxx** [comment]
I don't like this new Point Filter in 2021.2.0_2 version. Want to use Opentrack with Accela only, as before but with new fps on camera.
PT makes view slow, with huge delay, i hate it. Ideal view for me is like human eyes must be fast and precise, easy to stop in place you need without wobbling. Tuned Accela do it well.

**Reply by Len62** [reply]
Just disable the point filter?

**Reply by animusVoxxx** [reply]
2021.2.0_2 version not working without PT filter. lag, jittery movement, you know.

**Reply by Len62** [reply]
Ok, I'm still on 2021.2.0. Can't find the _2 version.

**Reply by animusVoxxx** [reply]
2021.2.0 is correct name. I have couple duplicate files in same folder.

**Reply by sthalik** [reply]
I disabled requesting grayscale from the ps3eye open driver. Does this build help you any? https://www.dropbox.com/s/6vhkrd2kmyfrpqw/opentrack-master-20220102_01.7z

And yes, Accela should function as it previously did without the point filter.

**Reply by animusVoxxx** [reply]
No changes for me with master version. Circles around diodes keep jittering with different settings.

**Reply by sthalik** [reply]
@animusVoxxx what are your PT thresholding settings? Try screenshotting the first PT tab.

**Comment by Vortex76** [comment]
Yep Thats it..!!! ( *opentrack-master-20220104_01.7z* )  Its pretty stable
using old accella settings, with threshold 9.73, pf 0.13, limit 0.32 seems
to work fine! I had to mess around a bit with the values to stabilise more
but now its smooth as before the PF was introduced.
One problem though is that for every different person will need that bit of
fine tuning so still you ll have lots of people complaining..Maybe write a
paragraph on what every setting does so its easier for the user to
comprehend and adjust to his needs? Just saying.. Anyway, I think you
nailed the problem. Thanks Stanislaw!!!!

**Reply by sthalik** [reply]
It should work just as well with Accela or Hamilton as before the point filter got introduced. Can you try out that configuration?

**Reply by Vortex76** [reply]
With Accella on and PF on is really fine as long as you tweak the settings. I use exaggerated values in Accela from before so I had to go easy on the PF settings. So : 
Accela ON + PF OFF = OK
Accela ON + PF ON = OK
Do you need me try some specific setup?

**Comment by sthalik** [comment]
Something weird is going on. Can you notice how the cyan ellipse drifts off the point's center when using extreme orientation? For me it points to mostly empty space on high yaw.

**Reply by Vortex76** [reply]
try pf 0.13 and limit 032. That stabilised it a lot for me. The limit mostly

**Reply by Vortex76** [reply]
![acc](https://user-images.githubusercontent.com/96124211/148293763-7012a089-339e-429e-b0d4-162b8a4f2952.PNG)
![PF](https://user-images.githubusercontent.com/96124211/148293766-8396b297-7971-4f22-8af0-fbbd762a6d37.PNG)

**Comment by sthalik** [comment]
@Vortex76 @animusVoxxx @dglava: Check this build: <<https://www.dropbox.com/s/42f5v14amkmpz40/opentrack-master-20220105_01.7z>>. I plan to release it at this state for 2022.1.0.

**Reply by Vortex76** [reply]
Thumbs up from my end on this build !

**Reply by sthalik** [reply]
Alright, so this is a bit confusing. I reverted two separate changes because the thing making the difference seems to be 4b77e13d86135b55bd33de49322ed42219680162. Please see whether this build works fine: https://www.dropbox.com/s/ik19faukb7vfzbz/opentrack-master-20220105_02.7z.

For reference, this is what was in _01 but got reverted in _02:
- call centroid estimation code back from 2012 prior to passing blobs to meanshift
- also use pixels below the threshold to estimate centroids

Now, in _02:
- more descriptive color options / make hardware grayscale optional
- don't let meanshift use pixels from beyond the rectangle where the blob is detected
- don't set a minimum value for auto thresholding (was set to brightness of 64, now is 1)

Basically only changes that are known to work should be made. What I did previously is make a ton of changes and see what sticks.

**Reply by 3itZ** [reply]
20220105_02 slight twitching is present. 
20220105_01 there is no such thing, everything is smooth

**Reply by sthalik** [reply]
I brought back the threshold and meanshift changes. This is the same as _01, but also has the freetrack fix. https://www.dropbox.com/s/fggp7823bluktbp/opentrack-master-20220105_04.7z

**Reply by Vortex76** [reply]
No problems for me with either 02 or 04, using the settings I posted above. They both seem to work really fine.

**Reply by 3itZ** [reply]
20220105_04 seemed to work well, there was little time for the test...

**Reply by animusVoxxx** [reply]
20220105_04 is better, now playable without point firter. Stepping a bit.

**Comment by Len62** [comment]
No problems with 02 or 04. Tried Vortex76 settings, but found some slight stutter. Reverted back and feels really good.

**Comment by Perfectionman** [comment]
After closing OpenTrack I lose all settings and when I relaunch it everything is back to default settings, NOTHING saved. I have done uninstalls and made sure to delete all files and reinstalled.  I have disabled OneDrive sync for the Documents directory.  I have tried the portable version and an older versions but same result.  I'm at a complete loss as for what to do :(

**Comment by sthalik** [comment]
I'm telling you, the sync hasn't been fully disabled. You probably have temporary files in `Documents/opentrack-2.3` and that signifies that they couldn't be renamed. They're not present otherwise.

Anyway run the app through Windows Console by dragging the executable onto the window, then press Enter. You should get an error in the form of,

```
    qDebug() << "error with .ini file" << settings.fileName() << settings.status();
```
